plugins/parsers/statsd/parser.go

```diff
@@ -0,0 +1,272 @@
+					continue
+				}
+			} else {
+				log.Printf(errmsg, "", line)
+				continue
+			}
+		}
+
+		var mtype string
+		// Validate metric type
+		switch pipesplit[1] {
+		case "g", "c", "s", "ms", "h":
+			mtype = pipesplit[1]
+		default:
+			log.Printf("E! Error: Statsd Metric type %s unsupported\n", pipesplit[1])
+			return nil, errors.New("Error Parsing statsd line")
+		}
+
+		v, err := parseFieldValue(mtype, pipesplit[0])
+		if err != nil {
+			return nil, err
+		}
+
+		// samplerate with counter
+		if samplerate > 0 && mtype == "c" {
+			v = int64(float64(v.(int64)) / samplerate)
+		}
+
+		fields := map[string]interface{}{
+			defaultFieldName: v,
+		}
+		tags[defaultTagKey] = mtype
+
+		m, err := metric.New(name, tags, fields, time.Now())
+		if err != nil {
+			return nil, err
+		}
+		metrics = append(metrics, m)
+
+		// samplerate with timing
+		if samplerate > 0 && (mtype == "ms" || mtype == "h") {
+			size := int(1.0/samplerate) - 1
+			if size > 0 {
+				sampleMetrics := make([]telegraf.Metric, size)
+				for i := 0; i < size; i++ {
+					sampleMetrics[i], err = metric.New(name, tags, fields, time.Now())
+					if err != nil {
+						return nil, err
+					}
+				}
+				metrics = append(metrics, sampleMetrics...)
+			}
+		}
+	}
+	return metrics, nil
+}
+
+func (p *Parser) parseName(bucket string) (string, string, map[string]string) {
+	tags := make(map[string]string)
+
+	bucketparts := strings.Split(bucket, ",")
+	// Parse out any tags in the bucket
+	if len(bucketparts) > 1 {
+		for _, btag := range bucketparts[1:] {
+			k, v := parseKeyValue(btag)
+			if k != "" {
+				tags[k] = v
+			}
+		}
+	}
+
+	var field string
+	name := bucketparts[0]
+
+	graphiteParser := p.graphiteParser
+	var err error
+
+	if graphiteParser == nil {
+		graphiteParser, err = graphite.NewGraphiteParser(
+			p.Separator, p.Templates, nil)
+		p.graphiteParser = graphiteParser
+	}
+
+	if err == nil {
+		graphiteParser.DefaultTags = tags
+		name, tags, field, _ = graphiteParser.ApplyTemplate(name)
+	}
+
+	if field == "" {
+		field = defaultFieldName
+	}
+
+	return name, field, tags
+}
+
+// Parse the key,value out of a string that looks like "key=value"
+func parseKeyValue(keyvalue string) (string, string) {
+	var key, val string
+
+	split := strings.Split(keyvalue, "=")
+	// Must be exactly 2 to get anything meaningful out of them
+	if len(split) == 2 {
+		key = split[0]
+		val = split[1]
+	} else if len(split) == 1 {
+		// why?
+		val = split[0]
+	}
+
+	return key, val
+}
+
+func parseFieldValue(mtype string, value string) (interface{}, error) {
+	switch mtype {
+	case "ms", "h":
+		v, err := strconv.ParseFloat(value, 64)
+		if err != nil {
+			return nil, err
+		}
+		return v, nil
+	case "c":
+		v, err := strconv.ParseInt(value, 10, 64)
+		if err != nil {
+			v2, err2 := strconv.ParseFloat(value, 64)
+			if err2 != nil {
+				return nil, errors.New("Error Parsing statsd line")
+			}
+			v = int64(v2)
+		}
+		return v, nil
+	case "s", "g":
```

Comment:
I think we should parse gauge as a float64, but instead add an additional tag if it is a relative value:
```
foo,metric_type=gauge,operation=+ value=42
```

Reply:
- Change type to float64.
- use "additive" may be better? 

`foo:50|g` --> `foo,metric_type=gauge value=50`
`foo:+50|g` --> `foo,metric_type=gauge,operation=additive value=50`
`foo:-50|g` --> `foo,metric_type=gauge,operation=additive value=-50`

Reply:
Sure, and subtractive for `-`?

Reply:
Oh nevermind, yeah your example is fine.